controllers/topicPage/TopicPage.js

```diff
@@ -18,43 +18,70 @@ class TopicPage {
     this.getTopicPageById = this.getTopicPageById.bind(this);
   }
 
+  async getFeedTopicPages(payload) {
+    const topicPages = await this._getStreamService.getTopicPages(
+      payload.id,
+      payload.limit,
+      payload.offset
+    );
+    let newTopicPagesWithBlock = new BlockServices(
+      payload.listBlockUser,
+      payload.listBlockDomain,
+      payload.listPostAnonymous
+    ).getHasBlock(topicPages);
+    let results = await filterFeeds(payload.userId, newTopicPagesWithBlock, payload.id);
+    return results;
+  }
+
+  async modifiedFeed(userId, post) {
+    if (post.verb === POST_VERB_POLL) {
+      let postPoll = await modifyPollPostObject(userId, post);
+      return postPoll;
+    } else {
+      return post;
+    }
+  }
+
   async getTopicPages(req, res) {
     let {id} = req.params;
     let {limit = MAX_FEED_FETCH_LIMIT, offset = 0} = req.query;
 
     try {
       this._validator.validateGetTopicPages({id});
-      const topicPages = await this._getStreamService.getTopicPages(id, limit, offset);
-
       const listBlockUser = await getListBlockUser(req.userId);
       const listBlockDomain = await getBlockDomain(req.userId);
       const listPostAnonymous = await getListBlockPostAnonymous(req.userId);
-      let newTopicPagesWithBlock = new BlockServices(
+
+      let data = [];
+      let payload = {
         listBlockUser,
         listBlockDomain,
-        listPostAnonymous
-      ).getHasBlock(topicPages);
-      let data = [];
-
-      newTopicPagesWithBlock = await filterFeeds(req.userId, newTopicPagesWithBlock);
-
-      for (let post of newTopicPagesWithBlock) {
-        if (post?.is_hide) continue;
+        listPostAnonymous,
+        id,
+        limit,
+        offset,
+        userId: req.userId
+      };
+      let fetch = 0;
+      while (data.length < limit && fetch < MAX_FEED_FETCH_LIMIT) {
+        if (fetch !== 0) {
+          payload.offset = parseInt(payload.offset) + parseInt(limit);
+        }
+        let newTopicPagesWithBlock = await this.getFeedTopicPages(payload);
 
-        if (post.verb === POST_VERB_POLL) {
-          let postPoll = await modifyPollPostObject(req.userId, post);
-          data.push(postPoll);
-        } else {
-          data.push(post);
+        for (let post of newTopicPagesWithBlock) {
+          if (post?.is_hide) continue;
+          data.push(await this.modifiedFeed(req.userId, post));
         }
+        fetch += 1;
       }
 
       res.status(200).json({
         code: 200,
         status: 'success',
         message: 'Success get topic pages',
         data: data || [],
-        offset: parseInt(offset) + parseInt(topicPages.length)
+        offset: parseInt(offset)
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

In the `getTopicPages` method, you've added a loop to fetch topic pages in batches until a specified limit is reached. This is a good approach for handling large amounts of data. However, there's a potential issue with the way you're updating the offset in the payload. You're adding the limit to the offset after each iteration, which could lead to skipping some pages if the number of pages fetched in a batch is less than the limit. Instead, you should add the actual number of pages fetched in the current iteration to the offset.

```diff
      while (data.length < limit && fetch < MAX_FEED_FETCH_LIMIT) {
        if (fetch !== 0) {
-         payload.offset = parseInt(payload.offset) + parseInt(limit);
+         payload.offset = parseInt(payload.offset) + newTopicPagesWithBlock.length;
        }
        let newTopicPagesWithBlock = await this.getFeedTopicPages(payload);
 
        for (let post of newTopicPagesWithBlock) {
          if (post?.is_hide) continue;
          data.push(await this.modifiedFeed(req.userId, post));
         }
        fetch += 1;
       }
```

Also, it's a good practice to handle errors at the lowest level possible. In this case, you should handle any exceptions that might occur when calling `this.getFeedTopicPages` and `this.modifiedFeed`.

```diff
        let newTopicPagesWithBlock;
        try {
          newTopicPagesWithBlock = await this.getFeedTopicPages(payload);
        } catch (error) {
          console.error('Error getting feed topic pages:', error);
          throw error;
        }
 
        for (let post of newTopicPagesWithBlock) {
          if (post?.is_hide) continue;
          let modifiedPost;
          try {
            modifiedPost = await this.modifiedFeed(req.userId, post);
          } catch (error) {
            console.error('Error modifying feed:', error);
            throw error;
          }
          data.push(modifiedPost);
         }
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->